### PR TITLE
06716 MAX_FULL_REHASHING_BUFFER_TIMEOUT is increased to 60 seconds.

### LINKED ISSUE #6716: Increase MAX_FULL_REHASHING_BUFFER_TIMEOUT to 60 seconds
### Problem

Current `MAX_FULL_REHASHING_BUFFER_TIMEOUT` equals to 10 seconds. Performance testing demonstrated that it's not enough:

```
2023-05-22 17:51:13.291 4068     INFO  STARTUP          <main> VirtualRootNode: Doing full rehash for the path range: 208748393 - 417496786  in the VirtualMap at [0 -> 11 -> 1]
2023-05-22 17:52:14.068 4562     ERROR EXCEPTION        <main> SwirldsPlatform: Saved state not loaded:
com.swirlds.common.merkle.synchronization.utility.MerkleSynchronizationException: Timed out waiting to supply a new leaf to the hashing iterator buffer
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.fullLeafRehash(VirtualRootNode.java:489) ~[swirlds-virtualmap-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.virtualmap.VirtualMap.fullLeafRehash(VirtualMap.java:231) ~[swirlds-virtualmap-0.39.0-adhoc.xa92eb106.jar:?]
	at com.hedera.node.app.service.mono.ServicesState.internalInit(ServicesState.java:410) ~[hedera-mono-service-0.39.0-06625-perfnet-testing.xb6b583f.jar:?]
	at com.hedera.node.app.service.mono.ServicesState.deserializedInit(ServicesState.java:312) ~[hedera-mono-service-0.39.0-06625-perfnet-testing.xb6b583f.jar:?]
	at com.hedera.node.app.service.mono.ServicesState.init(ServicesState.java:252) ~[hedera-mono-service-0.39.0-06625-perfnet-testing.xb6b583f.jar:?]
	at com.swirlds.platform.SwirldsPlatform.loadSavedState(SwirldsPlatform.java:973) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.SwirldsPlatform.initializeLoadedStateFromSignedState(SwirldsPlatform.java:934) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.SwirldsPlatform.<init>(SwirldsPlatform.java:613) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.createLocalPlatforms(Browser.java:689) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.startPlatforms(Browser.java:829) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.<init>(Browser.java:311) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.launch(Browser.java:541) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.parseCommandLineArgsAndLaunch(Browser.java:497) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.main(Browser.java:853) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
``` 

### Solution

Increase `MAX_FULL_REHASHING_BUFFER_TIMEOUT` to 60 seconds. I chose 60 seconds because we have similar hashing mechanism for reconnect. It has the same buffer size (10,000,000) and timeout of 60 seconds. My reasoning is that if it's enough for the hashing during reconnect, it should be enough for rehashing of 10m leaves as well. 

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by imalygin**
**Description**:
Current MAX_FULL_REHASHING_BUFFER_TIMEOUT equals to 10 seconds. Performance testing demonstrated that it's not enough. Increase MAX_FULL_REHASHING_BUFFER_TIMEOUT to 60 seconds. I chose 60 seconds because we have a similar hashing mechanism for the reconnect. It has the same buffer size (10,000,000) and timeout of 60 seconds. My reasoning is that if it's enough for the hashing during reconnect, it should be enough for rehashing of 10m leaves as well.

**Update:** Oleg suggested following improvements:
- instead of having a timeout on iterator, let's fail the whole rehashing if it can't make it on time. The timeout is set to 6 hours.
- to rehash a prod-sized virtual map it may take up to 3-4 hours. It necessary to see if it's making any progress. I added progress-tracking logs/

**Related issue(s)**:

Fixes #6716 , #5825 

**Notes for reviewer**:

Logs from perf test

```
2023-05-22 17:51:13.291 4068     INFO  STARTUP          <main> VirtualRootNode: Doing full rehash for the path range: 208748393 - 417496786  in the VirtualMap at [0 -> 11 -> 1]
2023-05-22 17:52:14.068 4562     ERROR EXCEPTION        <main> SwirldsPlatform: Saved state not loaded:
com.swirlds.common.merkle.synchronization.utility.MerkleSynchronizationException: Timed out waiting to supply a new leaf to the hashing iterator buffer
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.fullLeafRehash(VirtualRootNode.java:489) ~[swirlds-virtualmap-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.virtualmap.VirtualMap.fullLeafRehash(VirtualMap.java:231) ~[swirlds-virtualmap-0.39.0-adhoc.xa92eb106.jar:?]
	at com.hedera.node.app.service.mono.ServicesState.internalInit(ServicesState.java:410) ~[hedera-mono-service-0.39.0-06625-perfnet-testing.xb6b583f.jar:?]
	at com.hedera.node.app.service.mono.ServicesState.deserializedInit(ServicesState.java:312) ~[hedera-mono-service-0.39.0-06625-perfnet-testing.xb6b583f.jar:?]
	at com.hedera.node.app.service.mono.ServicesState.init(ServicesState.java:252) ~[hedera-mono-service-0.39.0-06625-perfnet-testing.xb6b583f.jar:?]
	at com.swirlds.platform.SwirldsPlatform.loadSavedState(SwirldsPlatform.java:973) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.SwirldsPlatform.initializeLoadedStateFromSignedState(SwirldsPlatform.java:934) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.SwirldsPlatform.<init>(SwirldsPlatform.java:613) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.createLocalPlatforms(Browser.java:689) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.startPlatforms(Browser.java:829) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.<init>(Browser.java:311) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.launch(Browser.java:541) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.parseCommandLineArgsAndLaunch(Browser.java:497) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
	at com.swirlds.platform.Browser.main(Browser.java:853) ~[swirlds-platform-core-0.39.0-adhoc.xa92eb106.jar:?]
```

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by artemananiev — DISMISSED** [review]

**Review by OlegMazurov — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java [hunk 1]
```diff
@@ -477,42 +479,80 @@ public void fullLeafRehash() {
...
-                                rehashIterator.supply(leafRecord, MAX_FULL_REHASHING_BUFFER_TIMEOUT, SECONDS);
-                        if (!success) {
-                            throw new MerkleSynchronizationException(
-                                    "Timed out waiting to supply a new leaf to the hashing iterator buffer");
+        // This background thread will be responsible for feeding the iterator with data.
+        new ThreadConfiguration(getStaticThreadManager())
+                .setComponent("virtualmap")
+                .setThreadName("leafFeeder")
+                .setRunnable(() -> {
+                    long onePercent = (lastLeafPath - firstLeafPath) / 100 + 1;
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:487]: final

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java [hunk 2]
```diff
@@ -477,42 +479,80 @@ public void fullLeafRehash() {
...
+                    for (long i = firstLeafPath; i <= lastLeafPath; i++) {
+                        try {
+                            VirtualLeafRecord<K, V> leafRecord = dataSource.loadLeafRecord(i);
+                            if (leafRecord != null) {
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:491]: Can it actually be null? To me, all records in [firstLeafPath, lastLeafPath] must be resolvable at the data source level. An assertion (or even better, exception) would be beneficial here

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java [hunk 3]
```diff
@@ -477,42 +479,80 @@ public void fullLeafRehash() {
...
-                    } catch (final Exception e) {
-                        throw new MerkleSynchronizationException("Failed to handle a leaf during full rehashing", e);
                     }
-                }
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
-            }
-        }
-        rehashIterator.close();
+                    rehashIterator.close();
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:523]: Should it be wrapped into `finally` block?

**Review by artemananiev — APPROVED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java (hunk 2 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java]: sounds good, I'll replace it with an assert.

**Review by OlegMazurov — APPROVED** [review]

**Pull request merged**